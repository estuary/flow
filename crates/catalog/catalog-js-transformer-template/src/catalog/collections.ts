// This file contains placeholder examples of collection types,
// which are over-written during catalog build.

export type ExampleSourceCollection = {
    hello: string;
};

export type ExampleDerivedCollection = {
    world: string;
    value: number;
};
