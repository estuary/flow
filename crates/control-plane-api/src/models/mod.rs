pub mod activation_status;
pub use self::activation_status::ActivationStatus;
pub mod affected_consumer;
pub use self::affected_consumer::AffectedConsumer;
pub mod api_error;
pub use self::api_error::ApiError;
pub mod auto_discover_failure;
pub use self::auto_discover_failure::AutoDiscoverFailure;
pub mod auto_discover_outcome;
pub use self::auto_discover_outcome::AutoDiscoverOutcome;
pub mod auto_discover_status;
pub use self::auto_discover_status::AutoDiscoverStatus;
pub mod catalog_type;
pub use self::catalog_type::CatalogType;
pub mod discover_change;
pub use self::discover_change::DiscoverChange;
pub mod error;
pub use self::error::Error;
pub mod evolved_collection;
pub use self::evolved_collection::EvolvedCollection;
pub mod incompatible_collection;
pub use self::incompatible_collection::IncompatibleCollection;
pub mod inferred_schema_status;
pub use self::inferred_schema_status::InferredSchemaStatus;
pub mod job_status;
pub use self::job_status::JobStatus;
pub mod job_status_one_of;
pub use self::job_status_one_of::JobStatusOneOf;
pub mod job_status_one_of_1;
pub use self::job_status_one_of_1::JobStatusOneOf1;
pub mod job_status_one_of_2;
pub use self::job_status_one_of_2::JobStatusOneOf2;
pub mod job_status_one_of_3;
pub use self::job_status_one_of_3::JobStatusOneOf3;
pub mod job_status_one_of_4;
pub use self::job_status_one_of_4::JobStatusOneOf4;
pub mod job_status_one_of_5;
pub use self::job_status_one_of_5::JobStatusOneOf5;
pub mod job_status_one_of_6;
pub use self::job_status_one_of_6::JobStatusOneOf6;
pub mod job_status_one_of_7;
pub use self::job_status_one_of_7::JobStatusOneOf7;
pub mod job_status_one_of_8;
pub use self::job_status_one_of_8::JobStatusOneOf8;
pub mod lock_failure;
pub use self::lock_failure::LockFailure;
pub mod publication_info;
pub use self::publication_info::PublicationInfo;
pub mod publication_status;
pub use self::publication_status::PublicationStatus;
pub mod re_create_reason;
pub use self::re_create_reason::ReCreateReason;
pub mod rejected_field;
pub use self::rejected_field::RejectedField;
pub mod source_capture_status;
pub use self::source_capture_status::SourceCaptureStatus;
pub mod status_response;
pub use self::status_response::StatusResponse;
pub mod status_response_controller_status;
pub use self::status_response_controller_status::StatusResponseControllerStatus;
pub mod status_response_controller_status_one_of;
pub use self::status_response_controller_status_one_of::StatusResponseControllerStatusOneOf;
pub mod status_response_controller_status_one_of_1;
pub use self::status_response_controller_status_one_of_1::StatusResponseControllerStatusOneOf1;
pub mod status_response_controller_status_one_of_2;
pub use self::status_response_controller_status_one_of_2::StatusResponseControllerStatusOneOf2;
pub mod status_response_controller_status_one_of_3;
pub use self::status_response_controller_status_one_of_3::StatusResponseControllerStatusOneOf3;
