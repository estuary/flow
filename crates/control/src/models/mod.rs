use std::convert::TryInto;
use std::fmt::Display;
use std::str::FromStr;

use base64::display::Base64Display;

pub mod accounts;
pub mod connector_images;
pub mod connectors;
pub mod credentials;

#[derive(Debug, thiserror::Error)]
pub enum MalformedIdError {
    #[error("id does not appear to be valid")]
    Decode,
    #[error("id with an unexpected length")]
    Length,
}

static ENCODING_CONFIG: base64::Config = base64::URL_SAFE_NO_PAD;

/// `Id` is a transparent wrapper around our non-sequential Postgres keys. We
/// wish to expose these Id values as base64 encoded values. This wrapper
/// handles all these serialization and deserialization concerns.
///
/// When creating a new Postgres-backed record or relationship, it is expected
/// to use `Id` for primary and foreign keys.
#[serde_as]
#[derive(
    Clone,
    Copy,
    Debug,
    Eq,
    Ord,
    PartialEq,
    PartialOrd,
    DeserializeFromStr,
    SerializeDisplay,
    sqlx::Type,
)]
#[sqlx(transparent)]
pub struct Id(i64);

impl Id {
    pub const fn new(inner: i64) -> Self {
        Self(inner)
    }

    /// Retrieves the raw value of the Id. Not to be directly exposed to
    /// customers, but not a secret.
    pub fn to_i64(&self) -> i64 {
        self.0
    }

    /// Converts the Id value to bytes. **Must** be the inverse operation of `from_bytes`.
    pub(crate) fn bytes(&self) -> [u8; 8] {
        self.0.to_be_bytes()
    }

    /// Converts bytes into an Id value. **Must** be the inverse operation of `bytes`.
    fn from_bytes(bytes: [u8; 8]) -> Self {
        Id::new(i64::from_be_bytes(bytes))
    }

    /// An Id that does not correspond to a persisted resource, but rather is a
    /// one-time use identifier. Nonces must not be reused.
    pub fn nonce() -> Self {
        Id::new(rand::random())
    }
}

impl Display for Id {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            Base64Display::with_config(&self.bytes(), ENCODING_CONFIG)
        )
    }
}

impl FromStr for Id {
    type Err = MalformedIdError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let bytes =
            base64::decode_config(s, ENCODING_CONFIG).map_err(|_| MalformedIdError::Decode)?;
        let bytes = bytes.try_into().map_err(|_| MalformedIdError::Length)?;
        Ok(Id::from_bytes(bytes))
    }
}

impl From<Id> for String {
    fn from(id: Id) -> Self {
        id.to_string()
    }
}
