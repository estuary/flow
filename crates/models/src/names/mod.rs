use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use serde_json::{from_value, json};

mod collection;
mod journal;
mod reference;
mod schema_support;
mod shards;

pub use collection::PartitionSelector;
pub use journal::{
    BucketType, CompressionCodec, FragmentTemplate, JournalTemplate, StorageMapping, Store,
};
pub use reference::{
    Capture, Collection, CompositeKey, Field, JsonPointer, Materialization, Object, Prefix, Rule,
    Test, Transform,
};
pub use shards::ShardTemplate;

fn duration_schema(_: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
    from_value(json!({
        "type": ["string", "null"],
        "pattern": "^\\d+(s|m|h)$"
    }))
    .unwrap()
}

fn is_false(b: &bool) -> bool {
    !*b
}

/// Lambdas are user functions which are invoked by the Flow runtime to
/// process and transform source collection documents into derived collections.
/// Flow supports multiple lambda run-times, with a current focus on TypeScript
/// and remote HTTP APIs.
///
/// TypeScript lambdas are invoked within on-demand run-times, which are
/// automatically started and scaled by Flow's task distribution in order
/// to best co-locate data and processing, as well as to manage fail-over.
///
/// Remote lambdas may be called from many Flow tasks, and are up to the
/// API provider to provision and scale.
#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
#[schemars(example = "Lambda::example_typescript")]
#[schemars(example = "Lambda::example_remote")]
pub enum Lambda {
    Typescript,
    Remote(String),
}
