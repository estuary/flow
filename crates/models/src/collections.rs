use super::{CompositeKey, Derivation, Field, Id, JournalTemplate, JsonPointer, Schema};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use serde_json::{from_value, json};
use std::collections::BTreeMap;

/// Collection describes a set of related documents, where each adheres to a
/// common schema and grouping key. Collections are append-only: once a document
/// is added to a collection, it is never removed. However, it may be replaced
/// or updated (either in whole, or in part) by a future document sharing its
/// key. Each new document of a given key is "reduced" into existing documents
/// of the key. By default, this reduction is achieved by completely replacing
/// the previous document, but much richer reduction behaviors can be specified
/// through the use of annotated reduction strategies of the collection schema.
#[derive(Serialize, Deserialize, Clone, Debug, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
#[schemars(example = "CollectionDef::example")]
pub struct CollectionDef {
    /// # Schema against which collection documents are validated and reduced on write and read.
    #[schemars(example = "Schema::example_relative")]
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub schema: Option<Schema>,
    /// # Schema against which collection documents are validated and reduced on write.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub write_schema: Option<Schema>,
    /// # Schema against which collection documents are validated and reduced on read.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub read_schema: Option<Schema>,
    /// # Composite key of this collection.
    pub key: CompositeKey,
    /// # Projections and logical partitions of this collection.
    #[schemars(schema_with = "projections_schema")]
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub projections: BTreeMap<Field, Projection>,
    /// # Template for journals of this collection.
    #[serde(default, skip_serializing_if = "JournalTemplate::is_empty")]
    pub journals: JournalTemplate,
    // # Derivation which builds this collection as transformations of other collections.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub derive: Option<Derivation>,
    /// # Expected publication ID of this collection within the control plane.
    /// When present, a publication of the collection will fail if the
    /// last publication ID in the control plane doesn't match this value.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expect_pub_id: Option<Id>,
    /// # Delete this collection within the control plane.
    /// When true, a publication will delete this collection.
    #[serde(default, skip_serializing_if = "super::is_false")]
    pub delete: bool,
}

impl CollectionDef {
    pub fn example() -> Self {
        Self {
            schema: Some(Schema::example_inline_basic()),
            write_schema: None,
            read_schema: None,
            key: CompositeKey::example(),
            projections: BTreeMap::new(),
            journals: JournalTemplate::default(),
            derive: None,
            expect_pub_id: None,
            delete: false,
        }
    }
}

/// Projections are named locations within a collection document which
/// may be used for logical partitioning or directly exposed to databases
/// into which collections are materialized.
#[derive(Serialize, Deserialize, Clone, Debug, JsonSchema)]
#[serde(untagged, deny_unknown_fields, rename_all = "camelCase")]
pub enum Projection {
    Pointer(JsonPointer),
    Extended {
        /// # Location of this projection.
        location: JsonPointer,
        /// # Is this projection a logical partition?
        #[serde(default)]
        partition: bool,
    },
}

impl Projection {
    pub fn as_parts(&self) -> (&JsonPointer, bool) {
        match self {
            Self::Pointer(location) => (location, false),
            Self::Extended {
                location,
                partition,
            } => (location, *partition),
        }
    }

    fn example_pointer() -> Self {
        Self::Pointer(JsonPointer::example())
    }

    fn example_extended() -> Self {
        Projection::Extended {
            location: JsonPointer::example(),
            partition: true,
        }
    }
}

fn projections_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
    let schema = Field::json_schema(gen);
    gen.definitions_mut().insert(Field::schema_name(), schema);

    let schema = Projection::json_schema(gen);
    gen.definitions_mut()
        .insert(Projection::schema_name(), schema);

    from_value(json!({
        "type": "object",
        "patternProperties": {
            Field::schema_pattern(): {
                "$ref": format!("#/definitions/{}", Projection::schema_name()),
            },
        },
        "additionalProperties": false,
        "examples": [{
            "a_field": Projection::example_pointer(),
            "a_partition": Projection::example_extended(),
        }],
    }))
    .unwrap()
}

impl super::ModelDef for CollectionDef {
    fn sources(&self) -> impl Iterator<Item = &crate::Source> {
        self.derive
            .iter()
            .map(|derive| derive.transforms.iter().map(|transform| &transform.source))
            .flatten()
    }
    fn targets(&self) -> impl Iterator<Item = &crate::Collection> {
        std::iter::empty()
    }

    fn catalog_type(&self) -> crate::CatalogType {
        crate::CatalogType::Collection
    }
}
