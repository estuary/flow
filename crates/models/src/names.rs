use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::BTreeMap;

// This module holds project-wide, type-safe wrappers, enums, and *very* simple
// structures which identify or name Flow concepts, and must be referenced from
// multiple different crates.

/// Collection names consist of Unicode letters, numbers, and symbols: - _ . /
///
/// Spaces and other special characters are disallowed.
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[schemars(example = "Collection::example")]
pub struct Collection(#[schemars(schema_with = "Collection::schema")] String);

impl Collection {
    pub fn new(name: impl Into<String>) -> Self {
        Self(name.into())
    }
    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl std::ops::Deref for Collection {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq, Eq, PartialOrd, Ord)]
#[schemars(example = "Transform::example")]
pub struct Transform(String);

impl Transform {
    pub fn new(name: impl Into<String>) -> Self {
        Self(name.into())
    }
    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl std::ops::Deref for Transform {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

/// Endpoint names a Flow endpoint.
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq, Eq, PartialOrd, Ord)]
#[schemars(example = "Endpoint::example")]
pub struct Endpoint(String);

impl Endpoint {
    pub fn new(name: impl Into<String>) -> Self {
        Self(name.into())
    }
}

impl std::ops::Deref for Endpoint {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

/// Materialization names a Flow materialization.
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq, Eq, PartialOrd, Ord)]
#[schemars(example = "Materialization::example")]
pub struct Materialization(String);

impl Materialization {
    pub fn new(name: impl Into<String>) -> Self {
        Self(name.into())
    }
    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl std::ops::Deref for Materialization {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

/// Capture names a Flow capture.
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq, Eq, PartialOrd, Ord)]
#[schemars(example = "Capture::example")]
pub struct Capture(String);

impl Capture {
    pub fn new(name: impl Into<String>) -> Self {
        Self(name.into())
    }
}

impl std::ops::Deref for Capture {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

/// Test names a Flow catalog test.
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq, Eq, PartialOrd, Ord)]
#[schemars(example = "Test::example")]
pub struct Test(String);

impl Test {
    pub fn new(name: impl Into<String>) -> Self {
        Self(name.into())
    }
}

impl std::ops::Deref for Test {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

/// Rule names a specification rule.
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq, Eq, PartialOrd, Ord)]
#[schemars(example = "Rule::example")]
pub struct Rule(String);

impl Rule {
    pub fn new(name: impl Into<String>) -> Self {
        Self(name.into())
    }
}

impl std::ops::Deref for Rule {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

/// JSON Pointer which identifies a location in a document.
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq, Eq, PartialOrd, Ord)]
#[schemars(example = "JsonPointer::example")]
pub struct JsonPointer(#[schemars(schema_with = "JsonPointer::schema")] String);

impl JsonPointer {
    pub fn new(ptr: impl Into<String>) -> Self {
        Self(ptr.into())
    }
}

impl std::ops::Deref for JsonPointer {
    type Target = str;

    fn deref(&self) -> &str {
        &self.0
    }
}

impl AsRef<str> for JsonPointer {
    fn as_ref(&self) -> &str {
        &self.0
    }
}

/// Ordered JSON-Pointers which define how a composite key may be extracted from
/// a collection document.
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
#[schemars(example = "CompositeKey::example")]
pub struct CompositeKey(Vec<JsonPointer>);

impl CompositeKey {
    pub fn new(parts: impl Into<Vec<JsonPointer>>) -> Self {
        Self(parts.into())
    }
    pub fn example() -> Self {
        CompositeKey(vec![JsonPointer::example()])
    }
}

impl std::ops::Deref for CompositeKey {
    type Target = Vec<JsonPointer>;

    fn deref(&self) -> &Vec<JsonPointer> {
        &self.0
    }
}

/// Object is an alias for a JSON object.
pub type Object = serde_json::Map<String, Value>;

/// Lambdas are user functions which are invoked by the Flow runtime to
/// process and transform source collection documents into derived collections.
/// Flow supports multiple lambda run-times, with a current focus on TypeScript
/// and remote HTTP APIs.
///
/// TypeScript lambdas are invoked within on-demand run-times, which are
/// automatically started and scaled by Flow's task distribution in order
/// to best co-locate data and processing, as well as to manage fail-over.
///
/// Remote lambdas may be called from many Flow tasks, and are up to the
/// API provider to provision and scale.
#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
#[schemars(example = "Lambda::example_typescript")]
#[schemars(example = "Lambda::example_remote")]
pub enum Lambda {
    Typescript,
    Remote(String),
}

/// Partition selectors identify a desired subset of the
/// available logical partitions of a collection.
#[derive(Serialize, Deserialize, Debug, JsonSchema, Clone)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
#[schemars(example = "PartitionSelector::example")]
pub struct PartitionSelector {
    /// Partition field names and corresponding values which must be matched
    /// from the Source collection. Only documents having one of the specified
    /// values across all specified partition names will be matched. For example,
    ///   source: [App, Web]
    ///   region: [APAC]
    /// would mean only documents of 'App' or 'Web' source and also occurring
    /// in the 'APAC' region will be processed.
    #[serde(default)]
    pub include: BTreeMap<String, Vec<Value>>,
    /// Partition field names and values which are excluded from the source
    /// collection. Any documents matching *any one* of the partition values
    /// will be excluded.
    #[serde(default)]
    pub exclude: BTreeMap<String, Vec<Value>>,
}
