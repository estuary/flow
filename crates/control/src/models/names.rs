use std::convert::Infallible;
use std::fmt::Display;
use std::ops::Deref;
use std::str::FromStr;

use models::collate::{collate, normalize};

/// The canonical entity name as it is understood in the data plane. This is a
/// unicode-normalized version of the user-provided name. This will be what
/// users understand as the entity's name.
///
/// eg. "acmeCo", "acmeCo/anvils"
#[serde_as]
#[derive(Clone, Debug, DeserializeFromStr, SerializeDisplay, sqlx::Type)]
#[sqlx(transparent)]
pub struct CatalogName(String);

impl CatalogName {
    pub fn new(name: &str) -> Self {
        Self(normalize(name.chars()).collect::<String>())
    }
}

impl FromStr for CatalogName {
    type Err = Infallible;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(Self::new(s))
    }
}

impl Display for CatalogName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        (&self.0[..]).fmt(f)
    }
}

impl Deref for CatalogName {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        self.0.as_ref()
    }
}

/// The representation of the entity name which will be used for enforcing
/// uniqueness and prefixing constraints. Entity names should not differ by
/// unicode constructor or by case. By unicode-normalizing and case-folding the
/// name we can reliably uniquely-index catalog entity names.
///
/// eg. "acmeco", "acmeco/anvils"
#[serde_as]
#[derive(Clone, Debug, DeserializeFromStr, SerializeDisplay, sqlx::Type)]
#[sqlx(transparent)]
pub struct UniqueName(String);

impl UniqueName {
    pub fn new(name: &str) -> Self {
        Self(collate(name.chars()).collect::<String>())
    }
}

impl FromStr for UniqueName {
    type Err = Infallible;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(Self::new(s))
    }
}

impl Display for UniqueName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        (&self.0[..]).fmt(f)
    }
}

impl Deref for UniqueName {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        self.0.as_ref()
    }
}
