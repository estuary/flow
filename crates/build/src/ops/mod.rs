//! The Flow runtime automatically publishes statistics and logs related to each task. It publishes
//! this data to Flow collections, so that users can create derivations and materializations of
//! that data. This module generates the Flow specs and schemas for these collections.
use models::names;
use protocol::flow::ContentType;
use serde_json::Value;
use std::collections::BTreeSet;
use url::Url;

struct GeneratedSchema {
    url: Url,
    content: &'static [u8],
}

macro_rules! gen_schema {
    ($name:expr) => {
        GeneratedSchema {
            content: &*include_bytes!($name),
            url: builtin_url($name),
        }
    };
}

/// Adds ops collections to the given partial built catalog. The tables will be modified in place
/// to add the resources required for the ops (logs and stats) collections.
pub fn generate_ops_collections(tables: &mut sources::Tables) {
    let tenants = all_tenant_names(tables);
    let shard_schema = gen_schema!("ops-shard-schema.json");
    let stats_schema = gen_schema!("ops-stats-schema.json");
    let log_schema = gen_schema!("ops-log-schema.json");

    for schema in &[&shard_schema, &stats_schema, &log_schema] {
        tables.resources.insert_row(
            schema.url.clone(),
            ContentType::JsonSchema,
            bytes::Bytes::from_static(schema.content),
        );
        let dom = serde_json::from_slice::<Value>(schema.content)
            .expect("failed to parse builtin schema");
        tables.schema_docs.insert_row(schema.url.clone(), dom);
    }

    let importers = tables
        .resources
        .iter()
        .filter(|r| r.content_type == ContentType::CatalogSpec)
        .map(|r| r.resource.clone())
        .collect::<Vec<_>>();
    for importer in importers {
        tables.imports.insert_row(
            ops_collection_resource_url(),
            importer,
            ops_collection_resource_url(),
        );
    }

    for tenant in tenants {
        let logs_collection_name = format!("ops/{}/logs", tenant);
        let stats_collection_name = format!("ops/{}/stats", tenant);
        add_ops_collection(
            logs_collection_name,
            log_schema.url.clone(),
            Some(names::JsonPointer::new("/ts")),
            tables,
        );
        add_ops_collection(
            stats_collection_name,
            stats_schema.url.clone(),
            None,
            tables,
        );
    }
}

fn ops_collection_resource_url() -> Url {
    builtin_url("ops/generated/collections")
}

fn add_ops_collection(
    name: String,
    schema_url: Url,
    add_key: Option<names::JsonPointer>,
    tables: &mut sources::Tables,
) {
    let scope = ops_collection_resource_url();

    let name = names::Collection::new(name);
    let mut key = vec![
        names::JsonPointer::new("/shard/name"),
        names::JsonPointer::new("/shard/keyBegin"),
        names::JsonPointer::new("/shard/rClockBegin"),
    ];
    if let Some(add) = add_key {
        key.push(add);
    }

    tables.collections.insert_row(
        scope.clone(),
        name.clone(),
        schema_url,
        names::CompositeKey::new(key),
    );

    // Ops collections are partitioned by shard, so that each shard has dedicated journals for
    // logs and stats. The names of these partition fields are important because partitions are
    // ordered lexicographically, and we want the partitions to be ordered from most general to
    // most specific (kind -> name -> rangeKeyBegin -> rangeRClockBegin). This corresponds to the
    // hierarchy that will be used for storing fragments.
    let projections = &[
        ("kind", "/shard/kind"),
        ("name", "/shard/name"),
        ("rangeKeyBegin", "/shard/keyBegin"),
        ("rangeRClockBegin", "/shard/rClockBegin"),
    ];
    for projection in projections {
        tables.projections.insert_row(
            scope.clone(),
            name.clone(),
            projection.0.to_string(),
            names::JsonPointer::new(projection.1),
            true,
            true,
        );
    }
}

fn all_tenant_names(tables: &sources::Tables) -> BTreeSet<String> {
    let mut tenants = BTreeSet::new();
    let captures = tables.captures.iter().map(|c| c.capture.as_str());
    let derivations = tables.derivations.iter().map(|d| d.derivation.as_str());
    let materializations = tables
        .materializations
        .iter()
        .map(|m| m.materialization.as_str());
    let iter = captures.chain(derivations).chain(materializations);
    for name in iter {
        let tenant = first_path_component(name);
        if !tenants.contains(tenant) {
            tenants.insert(tenant.to_string());
        }
    }
    tenants
}

fn first_path_component(task_name: &str) -> &str {
    match task_name.split_once('/') {
        Some((first, _)) => first,
        None => task_name,
    }
}

fn builtin_url(name: &str) -> Url {
    Url::parse(&format!("builtin://flow/{}", name)).unwrap()
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn ops_collections_are_generated() {
        use models::names;

        let mut tables = sources::Tables::default();
        tables.captures.insert_row(
            builtin_url("test-cap.flow.yaml#/collections/acmeCo~1foo"),
            names::Capture::new("acmeCo/foo"),
            protocol::flow::EndpointType::AirbyteSource,
            serde_json::json!({}),
            7u32,
        );
        tables.captures.insert_row(
            builtin_url("test-cap.flow.yaml#/collections/shamazon~1bar"),
            names::Capture::new("shamazon/bar"),
            protocol::flow::EndpointType::AirbyteSource,
            serde_json::json!({}),
            8u32,
        );
        tables.derivations.insert_row(
            builtin_url("test-der.flow.yaml#/collections/gooble~1ads"),
            names::Collection::new("gooble/ads"),
            builtin_url(
                "test-der.flow.yaml?ptr=/collections/shamazon~1bar/derivation/register/schema",
            ),
            Value::Null,
        );
        tables.derivations.insert_row(
            builtin_url("test-der.flow.yaml#/collections/acmeCo~1tnt"),
            names::Collection::new("acmeCo/tnt"),
            builtin_url(
                "test-der.flow.yaml?ptr=/collections/acmeCo~1tnt/derivation/register/schema",
            ),
            Value::Null,
        );
        tables.materializations.insert_row(
            builtin_url("test-mat.flow.yaml#/collections/justme"),
            names::Materialization::new("justme"),
            protocol::flow::EndpointType::Postgresql,
            Value::Null,
        );

        generate_ops_collections(&mut tables);

        insta::assert_debug_snapshot!("ops_generated_schema_docs", &tables.schema_docs);
        insta::assert_debug_snapshot!("ops_generated_collections", &tables.collections);
        insta::assert_debug_snapshot!("ops_generated_projections", &tables.projections);
    }
}
