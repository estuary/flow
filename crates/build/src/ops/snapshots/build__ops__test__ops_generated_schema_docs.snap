---
source: crates/build/src/ops/mod.rs
expression: "&tables.schema_docs"

---
[
    SchemaDoc {
        schema: builtin://flow/ops-log-schema.json,
        dom: {"$schema":"https://json-schema.org/draft-07/schema","description":"Logs related to the processing of a Flow capture, derivation, or materialization","properties":{"fields":{"additionalProperties":true,"description":"Map of keys and values that are associated with this log entry.","type":"object"},"level":{"enum":["debug","info","warn","error"]},"message":{"type":"string"},"shard":{"$ref":"ops-shard-schema.json"},"ts":{"description":"Timestamp corresponding to the start of the transaction","format":"date-time","type":"string"}},"required":["shard","ts","level","message"],"title":"Flow task logs","type":"object"},
    },
    SchemaDoc {
        schema: builtin://flow/ops-shard-schema.json,
        dom: {"$schema":"https://json-schema.org/draft-07/schema","description":"Identifies a specific shard of a task, which may be the source of a log message or metrics","properties":{"keyBegin":{"description":"The inclusive beginning of the shard's assigned key range","pattern":"[0-9A-F]{8}","type":"string"},"kind":{"description":"The type of the catalog task","enum":["capture","derivation","materialization"]},"name":{"description":"The name of the catalog task (without the task type prefix)","type":"string"},"rClockBegin":{"description":"The inclusive beginning of the shard's assigned rClock range","pattern":"[0-9A-F]{8}","type":"string"}},"required":["kind","name","keyBegin","rClockBegin"],"title":"Flow shard id","type":"object"},
    },
    SchemaDoc {
        schema: builtin://flow/ops-stats-schema.json,
        dom: {"$defs":{"combineMetrics":{"properties":{"bytes_total":{"description":"The number of bytes processed from or into the given collection capture or materialization. Note that for \ncaptures and materializations this relates to the size of Flow JSON documents not the representation used by the external system.\n","reduce":{"strategy":"sum"},"type":"integer"},"docs_total":{"description":"The number of documents processed from or into the given collection capture or materialization","reduce":{"strategy":"sum"},"type":"integer"}},"reduce":{"strategy":"merge"},"required":["docs_total","bytes_total"],"type":"object"},"lambdaMetrics":{"properties":{"duration_seconds":{"description":"The total number of seconds elapsed for all invocations.","reduce":{"strategy":"sum"},"type":"number"},"invocations_total":{"description":"The number of invocations during this transaction","reduce":{"strategy":"sum"},"type":"integer"}},"reduce":{"strategy":"merge"},"required":["invocations_total","duration_seconds"],"type":"object"}},"$schema":"https://json-schema.org/draft-07/schema","description":"Statistics related to the processing of a Flow capture, derivation, or materialization","properties":{"combine":{"properties":{"left":{"$ref":"#/$defs/combineMetrics"},"out":{"$ref":"#/$defs/combineMetrics"},"right":{"$ref":"#/$defs/combineMetrics"}},"reduce":{"strategy":"merge"},"type":"object"},"connector":{"properties":{"start":{"description":"Counter of connector invocations","reduce":{"strategy":"sum"},"type":"integer"}},"reduce":{"strategy":"merge"},"type":"object"},"lambda":{"properties":{"publish":{"$ref":"#/$defs/lambdaMetrics"},"update":{"$ref":"#/$defs/lambdaMetrics"}},"reduce":{"strategy":"merge"},"type":"object"},"shard":{"$ref":"ops-shard-schema.json"},"ts":{"description":"Timestamp corresponding to the start of the transaction","format":"date-time","reduce":{"strategy":"maximize"},"type":"string"}},"reduce":{"strategy":"merge"},"required":["shard","ts","combine"],"title":"Flow task stats","type":"object"},
    },
]
